### ISSUE TITLE
Crash - [SDWebImageDownloaderOperation start] Crashing with different reasons

### ISSUE THREAD

**Issue by clevelandlam** [rs/SDWebImage#1013]
Hello there, 

I've been encountering this crash occasionally, but cannot be reproduced intentionally. And there are different type of crashes at the same line - please kindly take a look and suggestions are welcomed.

Reason  - NSGenericException	 *** Collection <__NSSetM: 0x14536ff0> was mutated while being enumerated.

```
0	CoreFoundation 0x2ef55ecb __exceptionPreprocess + 131
1	libobjc.A.dylib 0x39a36ce7 objc_exception_throw + 36
2	CoreFoundation 0x2ef559b9 __NSFastEnumerationMutationHandler + 126
3	Foundation 0x2f8c1d31 -[NSISEngine variableToWorkOnAmongVariablesWithIntegralizationViolationsIgnoringLostCauses:varsAlreadyAdjusted:] + 246
4	Foundation 0x2f8c1a13 __44-[NSISEngine fixupIntegralizationViolations]_block_invoke + 896
5	Foundation 0x2f8bcd33 -[NSISEngine withBehaviors:performModifications:] + 208
6	Foundation 0x2f8c1675 -[NSISEngine fixupIntegralizationViolations] + 90
7	Foundation 0x2f8c1077 -[NSISEngine optimize] + 192
8	Foundation 0x2f8bcd9d -[NSISEngine withBehaviors:performModifications:] + 314
9	UIKit 0x318950c7 -[UIWindow(UIConstraintBasedLayout) updateConstraintsIfNeeded] + 216
10	UIKit 0x31803cff -[UIWindow(UIConstraintBasedLayout) layoutSublayersOfLayer:] + 84
11	QuartzCore 0x3140331b -[CALayer layoutSublayers] + 140
12	QuartzCore 0x313feb3f _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 348
13	QuartzCore 0x313fe9d1 _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 14
14	QuartzCore 0x313fe3e5 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 226
15	QuartzCore 0x313fe1f7 _ZN2CA11Transaction6commitEv + 312
16	QuartzCore 0x313f7f1d _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 54
17	CoreFoundation 0x2ef20ff9 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 18
18	CoreFoundation 0x2ef1e987 __CFRunLoopDoObservers + 284
19	CoreFoundation 0x2ee8974b CFRunLoopRunSpecific + 556
20	CoreFoundation 0x2eecd34b CFRunLoopRun + 96
21	Jtab 0x0018974b -[SDWebImageDownloaderOperation start] (SDWebImageDownloaderOperation.m:107)
22	Foundation 0x2f92e745 __NSOQSchedule_f + 58
23	libdispatch.dylib 0x39f2425f _dispatch_async_redirect_invoke + 108
24	libdispatch.dylib 0x39f25689 _dispatch_root_queue_drain + 226
25	libdispatch.dylib 0x39f258dd _dispatch_worker_thread2 + 54
26	libsystem_pthread.dylib 0x3a050c17 _pthread_wqthread + 296
27	libsystem_pthread.dylib 0x3a050adc start_wqthread + 6

SIGSEGV	
-[SDWebImageDownloaderOperation start] (SDWebImageDownloaderOperation.m:107)

0	CoreFoundation 0x2e78ccda CFDataGetLength + 6
1	Foundation 0x2f2ece53 NSISSparseVectorDescription + 96
2	Foundation 0x2f2f2071 -[NSISObjectiveLinearExpression description] + 266
3	Foundation 0x2f17b1d9 _NSDescriptionWithLocaleFunc + 86
4	CoreFoundation 0x2e833edb __CFStringAppendFormatCore + 5656
5	CoreFoundation 0x2e7a3611 _CFStringCreateWithFormatAndArgumentsAux + 74
6	CoreFoundation 0x2e7cbc03 _CFLogvEx + 84
7	Foundation 0x2f18a19f NSLogv + 92
8	Foundation 0x2f18a137 NSLog + 24
9	Foundation 0x2f1c92f1 -[NSISEngine minimizeConstantInObjectiveRowWithHead:] + 182
10	Foundation 0x2f1c8e57 -[NSISEngine optimize] + 172
11	Foundation 0x2f1c4b95 -[NSISEngine withBehaviors:performModifications:] + 314
12	UIKit 0x31188915 -[UIWindow(UIConstraintBasedLayout) updateConstraintsIfNeeded] + 218
13	UIKit 0x310f75eb -[UIWindow(UIConstraintBasedLayout) layoutSublayersOfLayer:] + 84
14	QuartzCore 0x30cf862b -[CALayer layoutSublayers] + 140
15	QuartzCore 0x30cf3e3b _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 348
16	QuartzCore 0x30cf3ccd _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 14
17	QuartzCore 0x30cf36df _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 228
18	QuartzCore 0x30cf34ef _ZN2CA11Transaction6commitEv + 312
19	QuartzCore 0x30ced21d _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 54
20	CoreFoundation 0x2e829255 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 18
21	CoreFoundation 0x2e826bf9 __CFRunLoopDoObservers + 282
22	CoreFoundation 0x2e791ee1 CFRunLoopRunSpecific + 554
23	CoreFoundation 0x2e7d59b3 CFRunLoopRun + 96
24	Jtab 0x0020374b -[SDWebImageDownloaderOperation start] (SDWebImageDownloaderOperation.m:107)
25	Foundation 0x2f23676d __NSOQSchedule_f + 58
26	libdispatch.dylib 0x39917293 _dispatch_async_redirect_invoke$VARIANT$mp + 108
27	libdispatch.dylib 0x39918ad7 _dispatch_root_queue_drain + 220
28	libdispatch.dylib 0x39918d29 _dispatch_worker_thread2 + 54
29	libsystem_pthread.dylib 0x39a53bd3 _pthread_wqthread + 296
30	libsystem_pthread.dylib 0x39a53a98 start_wqthread + 6
```

**Comment by flovilmart** [comment]
Doesn't look related to SDWebImage

**Comment by bpoplauschi** [comment]
Indeed, looks constraints related

**Comment by flovilmart** [comment]
@clevelandlam Can you provide the code that generated those crashes so we can further investigate?

**Comment by timkoma** [comment]
Same here, crashed thread: 12
So it seems that UIKit is called from backgroud thread.
 @clevelandlam which thread was it in your case?

**Comment by flovilmart** [comment]
@timkoma the thread number doesn't help much. Can you provide the backtrace? 

In the debugger, type bt and enter. That should print it

**Comment by flovilmart** [comment]
@cleverlandlam it seems you have an observer firing on the download operation start that leads to the crash. Without having your code, that seems pretty though to help you

**Comment by timkoma** [comment]
Exception Backtrace:
1	CoreFoundation	__exceptionPreprocess + 132
2	libobjc.A.dylib	objc_exception_throw + 56
3	CoreFoundation	__NSFastEnumerationMutationHandler + 136
4	Foundation	-[NSISEngine chooseOutgoingRowHeadForIncomingRowHead:] + 296
5	Foundation	-[NSISEngine minimizeConstantInObjectiveRowWithHead:] + 176
6	Foundation	-[NSISEngine tryUsingArtificialVariableToAddConstraintWithMarker:rowBody:usingInfeasibilityHandlingBehavior:mutuallyExclusiveConstraints:] + 284
7	Foundation	-[NSISEngine tryToAddConstraintWithMarker:expression:integralizationAdjustment:mutuallyExclusiveConstraints:] + 680
8	Foundation	-[NSLayoutConstraint _addLoweredExpression:toEngine:integralizationAdjustment:lastLoweredConstantWasRounded:mutuallyExclusiveConstraints:] + 316
9	Foundation	-[NSLayoutConstraint _addToEngine:integralizationAdjustment:mutuallyExclusiveConstraints:] + 256
10	UIKit	-[UIView(UIConstraintBasedLayout) _layoutEngine_didAddLayoutConstraint:roundingAdjustment:mutuallyExclusiveConstraints:] + 452
11	UIKit	-[UIView(UIConstraintBasedLayout) _constraints_didChangeAutoresizingConstraintsArrayForContainedView:] + 248
12	UIKit	-[UIView(UIConstraintBasedLayout) _updateAutoresizingConstraints] + 104
13	UIKit	-[UIView(AdditionalLayoutSupport) updateConstraints] + 124
14	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 244
15	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
16	CoreFoundation	CFArrayApplyFunction + 64
17	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
18	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
19	CoreFoundation	CFArrayApplyFunction + 64
20	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
21	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
22	CoreFoundation	CFArrayApplyFunction + 64
23	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
24	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
25	CoreFoundation	CFArrayApplyFunction + 64
26	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
27	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
28	CoreFoundation	CFArrayApplyFunction + 64
29	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
30	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
31	CoreFoundation	CFArrayApplyFunction + 64
32	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
33	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
34	CoreFoundation	CFArrayApplyFunction + 64
35	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
36	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
37	CoreFoundation	CFArrayApplyFunction + 64
38	UIKit	-[UIView(AdditionalLayoutSupport) _internalUpdateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 144
39	UIKit	-[UIView(AdditionalLayoutSupport) _updateConstraintsIfNeededAccumulatingViewsNeedingSecondPass:] + 140
40	Foundation	-[NSISEngine withBehaviors:performModifications:] + 132
41	UIKit	-[UIWindow(UIConstraintBasedLayout) updateConstraintsIfNeeded] + 224
42	UIKit	-[UIWindow(UIConstraintBasedLayout) layoutSublayersOfLayer:] + 88
43	QuartzCore	-[CALayer layoutSublayers] + 180
44	QuartzCore	CA::Layer::layout_if_needed(CA::Transaction*) + 296
45	QuartzCore	CA::Layer::layout_and_display_if_needed(CA::Transaction*) + 28
46	QuartzCore	CA::Context::commit_transaction(CA::Transaction*) + 276
47	QuartzCore	CA::Transaction::commit() + 420
48	QuartzCore	CA::Transaction::observer_callback(__CFRunLoopObserver*, unsigned long, void*) + 76
49	CoreFoundation	__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 28
50	CoreFoundation	__CFRunLoopDoObservers + 368
51	CoreFoundation	__CFRunLoopRun + 760
52	CoreFoundation	CFRunLoopRunSpecific + 448
53	CoreFoundation	CFRunLoopRun + 108
54	HomeTeam	-[SDWebImageDownloaderOperation start] (SDWebImageDownloaderOperation.m:107)
55	Foundation	__NSOQSchedule_f + 72
56	libdispatch.dylib	_dispatch_client_callout + 12
57	libdispatch.dylib	_dispatch_async_redirect_invoke + 148
58	libdispatch.dylib	_dispatch_client_callout + 12
59	libdispatch.dylib	_dispatch_root_queue_drain + 552
60	libdispatch.dylib	_dispatch_worker_thread2 + 72
61	libsystem_pthread.dylib	_pthread_wqthread + 352
62	libsystem_pthread.dylib	start_wqthread + 0

**Comment by flovilmart** [comment]
It still seems unrelated to SDWebImage. Without the context where you call the framework, it's nearly impossible to understand why it triggers a CATransaction commit and that your constraints or views are mutated.

 

Can you share the code responsible for the crash?

**Comment by timkoma** [comment]
I can't share the code, but why it starts from SDWebImageDownloaderOperation start?

**Comment by flovilmart** [comment]
This is why your code could be helpful.


Why can’t you share a piece of it?




Keep in mind that your error is likely to a misusage of SDWebImage, not a bug in the framework.

I’ll gladly help you to find it.

**Comment by timkoma** [comment]
Here it is:
    - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
    {
        HTBookPage * page = [self.book.pages objectAtIndex:indexPath.item];
        HTBookPageCell * pageCell = [collectionView dequeueReusableCellWithReuseIdentifier:@"BookPageCell" forIndexPath:indexPath];
    
        [pageCell.pageImage sd_setImageWithURL:[NSURL URLWithString:page.source]];
        return pageCell;
    }

**Comment by timkoma** [comment]
I think that it is related to SDWebImage because it starts from SDWebImageDownloaderOperation's start method

**Comment by flovilmart** [comment]
And seeing the code you provided, I can assure you that it should be properly working (I have the exact same calls). IMHO, something’s going on with the usage more than the framework itself.




Are all your layout constraints properly set? Do you have runtime disabled constraints?




The crash is related to the fact that there’s a relayout happening :


43 QuartzCore -[CALayer layoutSublayers] + 180





the relayout is not explicitly called by start] but by:





49 CoreFoundation __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION + 28





When the layoutSublayers occurs, something gets mutated along the way. Likely to be constraints:





13 UIKit -[UIView(AdditionalLayoutSupport) updateConstraints] + 124





And more precisely:





12 UIKit -[UIView(UIConstraintBasedLayout) _updateAutoresizingConstraints] + 104





My guts tells me that some constraints are broken in you collection view, cells or header view.

**Comment by pjchavarria** [comment]
I have the same issue, using Autolayout with UICollectionViewCells, crashes occasionally too. If i remove `sd_setImageWithURL` it won't crash.

**Comment by pjchavarria** [comment]
My problem was that i was calling collectionView.reloadData() from a background thread!

**Comment by jiehu5114** [comment]
@pjchavarria   did you have the demo?

**Comment by bpoplauschi** [comment]
Let's close this issue, see if it happens again with 4.0.0 (4.0.0-beta is available for download, 4.0.0 official release is on its way).
